File: sdk/go/alp_parser.go
package alpgo

import (
	"fmt"
	"strings"
)

type AlpParser struct{}

func (p *AlpParser) Parse(source string) ([]*AlpObject, error) {
	blocks := strings.Split(source, "\n\n")
	var objects []*AlpObject
	for _, block := range blocks {
		trimmed := strings.TrimSpace(block)
		if trimmed == "" {
			continue
		}
		obj, err := p.parseBlock(trimmed)
		if err != nil {
			return nil, err
		}
		if obj != nil {
			objects = append(objects, obj)
		}
	}
	return objects, nil
}

func (p *AlpParser) ParseSingle(source string) (*AlpObject, error) {
	trimmed := strings.TrimSpace(source)
	if trimmed == "" {
		return nil, fmt.Errorf("empty source provided to parser")
	}
	result, err := p.parseBlock(trimmed)
	if err != nil {
		return nil, err
	}
	if result == nil {
		return nil, fmt.Errorf("failed to parse ALP block")
	}
	return result, nil
}

func (p *AlpParser) parseBlock(block string) (*AlpObject, error) {
	lines := strings.Split(block, "\n")
	var id, objectType string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "id:") {
			id = strings.TrimSpace(trimmed[3:])
		} else if strings.HasPrefix(trimmed, "type:") {
			objectType = strings.TrimSpace(trimmed[5:])
		}
	}
	if id == "" || objectType == "" {
		return nil, nil
	}
	return NewAlpObject(id, objectType), nil
}

File: sdk/rust/src/alp_parser.rs
use crate::{AlpError, AlpObject};
use std::collections::HashMap;

pub struct AlpParser;

impl AlpParser {
    pub fn parse(source: &str) -> Result<Vec<AlpObject>, AlpError> {
        let blocks: Vec<&str> = source.split("\n\n").collect();
        let mut objects = Vec::new();
        for block in blocks {
            let trimmed = block.trim();
            if !trimmed.is_empty() {
                if let Some(obj) = Self::parse_block(trimmed)? {
                    objects.push(obj);
                }
            }
        }
        Ok(objects)
    }

    pub fn parse_single(source: &str) -> Result<AlpObject, AlpError> {
        let trimmed = source.trim();
        if trimmed.is_empty() {
            return Err(AlpError::new("Empty source provided to parser"));
        }
        let result = Self::parse_block(trimmed)?;
        result.ok_or_else(|| AlpError::new("Failed to parse ALP block"))
    }

    fn parse_block(block: &str) -> Result<Option<AlpObject>, AlpError> {
        let mut id = None;
        let mut object_type = None;
        for line in block.lines() {
            let trimmed = line.trim();
            if trimmed.starts_with("id:") {
                id = Some(trimmed[3..].trim().to_string());
            } else if trimmed.starts_with("type:") {
                object_type = Some(trimmed[5..].trim().to_string());
            }
        }
        match (id, object_type) {
            (Some(id), Some(object_type)) => Ok(Some(AlpObject::new(id, object_type))),
            _ => Ok(None),
        }
    }
}
